pkg/provisioner/ironic/ironic.go

```diff
@@ -794,6 +779,85 @@ func (p *ironicProvisioner) Deprovision(deleteIt bool) (result provisioner.Resul
 	}
 }
 
+// Delete removes the host from the provisioning system. It may be
+// called multiple times, and should return true for its dirty flag
+// until the deprovisioning operation is completed.
+func (p *ironicProvisioner) Delete() (result provisioner.Result, err error) {
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+	if ironicNode == nil {
+		p.log.Info("no node found, already deleted")
+		return result, nil
+	}
+
+	p.log.Info("deleting host",
+		"ID", ironicNode.UUID,
+		"lastError", ironicNode.LastError,
+		"current", ironicNode.ProvisionState,
+		"target", ironicNode.TargetProvisionState,
+		"deploy step", ironicNode.DeployStep,
+	)
+
+	if ironicNode.ProvisionState == nodes.Available {
+		// Move back to manageable so we can delete it cleanly.
```

Comment:
You can delete available nodes (it can be prevented, but it's allowed by default), there is no need for that.

Reply:
I got an error when I tried it, and this transition is easy enough to do.

Reply:
This is suspicious, something is going wrong. Available nodes can be deleted. Maybe you should make sure you don't keep `instance_uuid` on them when unprovisioning?

Reply:
@dhellmann did you want to follow up on @dtantsur 's last suggestion here, to see if the instace_uuid prevented deleting the node?

Reply:
I could do that. Will Ironic do something different during delete based on the state of the node?

Reply:
Note deletion us literally just deleting a record from the database (not to be confused with `deleted` action, which is an unfortunate name for unprovisioning).

Reply:
OK, that's pretty much what we want at this point. If the node is available, we haven't provisioned it so we don't need to clean it. And we want ironic not to argue with us when we say to delete it. I'll investigate whether removing the instance_uuid setting makes this possible without the state transition, but since this is working as it is I would like to proceed with this PR and do that other work separately.

Reply:
Looking into ironic's code, I see that destroy_node() in the conductor [checks a configuration option](https://opendev.org/openstack/ironic/src/branch/master/ironic/conductor/manager.py#L2274) before allowing Available nodes to be deleted. I think it's safer to keep the explicit state transition here.